src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -3532,6 +3541,156 @@ impl Coordinator {
         Ok((used_indexes, None, df_metainfo, transient_items))
     }
 
+    #[tracing::instrument(target = "optimizer", level = "trace", name = "optimize", skip_all)]
+    async fn explain_create_index_optimizer_pipeline(
+        &mut self,
+        name: QualifiedItemName,
+        index: Index,
+        broken: bool,
+    ) -> Result<
+        (
+            UsedIndexes,
+            Option<FastPathPlan>,
+            DataflowMetainfo,
+            BTreeMap<GlobalId, TransientItem>,
+        ),
+        AdapterError,
+    > {
+        use mz_repr::explain::trace_plan;
+
+        if broken {
+            tracing::warn!("EXPLAIN ... BROKEN <query> is known to leak memory, use with caution");
+        }
+
+        /// Like [`mz_ore::panic::catch_unwind`], with an extra guard that must be true
+        /// in order to wrap the function call in a [`mz_ore::panic::catch_unwind`] call.
+        fn catch_unwind<R, E, F>(guard: bool, stage: &'static str, f: F) -> Result<R, AdapterError>
+        where
+            F: FnOnce() -> Result<R, E>,
+            E: Into<AdapterError>,
+        {
+            if guard {
+                let r: Result<Result<R, E>, _> = mz_ore::panic::catch_unwind(AssertUnwindSafe(f));
```

Comment:
Could we have a test case that exercises this? Or does mz_panic not cause the panic in the right stage?

Reply:
I can only test `catch_unwind(broken, "global", f)` and `catch_unwind(broken, "local", f)` calls with `mz_panic()`. I'm not aware of a good way to test `hir_to_mir` and `finalize_dataflow` stages.

Reply:
I added some `EXPLAIN BROKEN <statement>` tests in a new commit.

Reply:
Removed the `EXPLAIN BROKEN <statement>` tests because they are breaking the CI analysis step (the `panic` and `ERROR` messages are flagged). If we have a fix for that and I'll add it back.

Reply:
I added the EXPLAIN BROKEN back with the CI analysis fix.